# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/quota/QuotaParent.cpp
# Commit: 396d9243d417
# Full Hash: 396d9243d4172aef258b24a09cf2a2574d86f87a
# Author: Jan Varga <jvarga@mozilla.com>
# Date: 2024-10-09 09:46:37
# Regressor Bug: 1866402
# File Overlap Count: 1
# Description:
#   Bug 1866402 - Add a dedicated method for clearing storages of a specific client; r=dom-storage-reviewers,jari
#   
#   Client storages can be currently cleared by using the optional argument of
#   nsIQuotaManagerService::ClearStoragesForPrincipal. However, supporting clearing
#   storages for both a principal and a client is not ideal for several reasons.
# ==============================================================================

diff -r fe4a050ac4da -r 396d9243d417 dom/quota/QuotaParent.cpp
--- a/dom/quota/QuotaParent.cpp	Wed Oct 09 05:34:35 2024 +0000
+++ b/dom/quota/QuotaParent.cpp	Wed Oct 09 05:34:36 2024 +0000
@@ -660,6 +660,40 @@
   return IPC_OK();
 }
 
+mozilla::ipc::IPCResult Quota::RecvClearStoragesForClient(
+    const Maybe<PersistenceType>& aPersistenceType,
+    const PrincipalInfo& aPrincipalInfo, const Type& aClientType,
+    ClearStoragesForClientResolver&& aResolver) {
+  AssertIsOnBackgroundThread();
+
+  QM_TRY(MOZ_TO_RESULT(!QuotaManager::IsShuttingDown()),
+         ResolveBoolResponseAndReturn(aResolver));
+
+  if (!TrustParams()) {
+    if (aPersistenceType) {
+      QM_TRY(MOZ_TO_RESULT(IsValidPersistenceType(*aPersistenceType)),
+             QM_CUF_AND_IPC_FAIL(this));
+    }
+
+    QM_TRY(MOZ_TO_RESULT(QuotaManager::IsPrincipalInfoValid(aPrincipalInfo)),
+           QM_CUF_AND_IPC_FAIL(this));
+
+    QM_TRY(MOZ_TO_RESULT(Client::IsValidType(aClientType)),
+           QM_CUF_AND_IPC_FAIL(this));
+  }
+
+  QM_TRY_UNWRAP(const NotNull<RefPtr<QuotaManager>> quotaManager,
+                QuotaManager::GetOrCreate(),
+                ResolveBoolResponseAndReturn(aResolver));
+
+  quotaManager
+      ->ClearStoragesForClient(aPersistenceType, aPrincipalInfo, aClientType)
+      ->Then(GetCurrentSerialEventTarget(), __func__,
+             BoolPromiseResolveOrRejectCallback(this, std::move(aResolver)));
+
+  return IPC_OK();
+}
+
 mozilla::ipc::IPCResult Quota::RecvClearStoragesForOriginPrefix(
     const Maybe<PersistenceType>& aPersistenceType,
     const PrincipalInfo& aPrincipalInfo,